### PR TITLE
Add ChainTrustValidationMode and CustomTrustStore properties to certauth.md

### LINKED ISSUE #15699: Add CustomTrustStore section to cert authN doc
See https://github.com/aspnet/AspNetCore/pull/15029#issuecomment-554140968

### PR DESCRIPTION
**Pull request by edwardneal**
Fixes #15699.

https://review.learn.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-8.0&branch=pr-en-us-31481#chaintrustvalidationmode

This adds information about the `ChainTrustValidationMode` and `CustomTrustStore` properties to the [Configure certificate authentication in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-7.0) article.

The two properties are designed to work hand-in-hand - `CustomTrustStore` is only used when `ChainTrustValidationMode` is set to `CustomRootTrust`. I've left them as two separate sections to keep it consistent with the rest of the properties.

`CustomTrustStore` also needs to contain every issuer in the client certificate chain: if an offline root certificate issues a per-tenant CA certificate and the per-tenant CA certificate issues a client certificate, this client certificate would only validate successfully if the property contained the offline root certificate and the per-tenant CA certificate. I've hopefully described this clearly enough in the new sections.

This is also the only way right now to allow ASP.NET to validate privately-trusted client certificates in Azure App Services (which is why I've included a reference to cloud environments.)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/security/authentication/certauth.md](https://github.com/dotnet/AspNetCore.Docs/blob/aef5fe127b0b7ae79880f3d74d6bc39e515dd7d2/aspnetcore/security/authentication/certauth.md) | [Configure certificate authentication in ASP.NET Core](https://review.learn.microsoft.com/en-us/aspnet/core/security/authentication/certauth?branch=pr-en-us-31481) |

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@Tratcher  please recommend a reviewer.

**Comment by Tratcher** [comment]
@blowdart

**Comment by blowdart** [comment]
Nope, whomever owns cert off now. @adityamandaleeka or @mkArtakMSFT ?

File: aspnetcore/security/authentication/certauth.md [hunk 1]
```diff
@@ -52,6 +52,22 @@ Default value: `CertificateTypes.Chained`
 
 This check validates that only the appropriate certificate type is allowed. If the app is using self-signed certificates, this option needs to be set to `CertificateTypes.All` or `CertificateTypes.SelfSigned`.
 
+### ChainTrustValidationMode
+
+Default value: `X509ChainTrustMode.System`
+
+The certificate presented by the client must chain to a trusted root certificate. This check allows you to control which trust store contains these root certificates.
+
+By default, the handler uses the system trust store. If the presented client certificate needs to chain to a root certificate which doesn't appear in the system trust store, (such as in a cloud environment) this option needs to be set to `X509ChainTrustMode.CustomRootTrust`.
```
Comment (halter73) [inline aspnetcore/security/authentication/certauth.md:61]:
```suggestion
By default, the handler uses the system trust store. If the presented client certificate needs to chain to a root certificate which doesn't appear in the system trust store, this option can be set to `X509ChainTrustMode.CustomRootTrust` to use the `CustomTrustStore` instead.
```

**Review by halter73 — APPROVED** [review]
LGTM. I'm not sure if you want more xrefs, but since all the nearby documentation doesn't use it, I assume it's fine.

File: aspnetcore/security/authentication/certauth.md [hunk 2]
```diff
@@ -52,6 +52,22 @@ Default value: `CertificateTypes.Chained`
 
 This check validates that only the appropriate certificate type is allowed. If the app is using self-signed certificates, this option needs to be set to `CertificateTypes.All` or `CertificateTypes.SelfSigned`.
 
+### ChainTrustValidationMode
+
+Default value: `X509ChainTrustMode.System`
```
Comment (Rick-Anderson) [inline aspnetcore/security/authentication/certauth.md:57]:
```suggestion
Default value: [X509ChainTrustMode.System](xref:System.Security.Cryptography.X509Certificates.X509ChainTrustMode.System)
```

File: aspnetcore/security/authentication/certauth.md [hunk 3]
```diff
@@ -52,6 +52,22 @@ Default value: `CertificateTypes.Chained`
...
+
+The certificate presented by the client must chain to a trusted root certificate. This check allows you to control which trust store contains these root certificates.
+
+By default, the handler uses the system trust store. If the presented client certificate needs to chain to a root certificate which doesn't appear in the system trust store, this option can be set to `X509ChainTrustMode.CustomRootTrust` to make the handler use the `CustomTrustStore` instead.
+
+### CustomTrustStore
+
+Default value: Empty `X509Certificate2Collection`
```
Comment (Rick-Anderson) [inline aspnetcore/security/authentication/certauth.md:65]:
```suggestion
Default value: Empty <xref:System.Security.Cryptography.X509Certificates.X509Certificate2Collection>
```

File: aspnetcore/security/authentication/certauth.md [hunk 4]
```diff
@@ -52,6 +52,22 @@ Default value: `CertificateTypes.Chained`
 
 This check validates that only the appropriate certificate type is allowed. If the app is using self-signed certificates, this option needs to be set to `CertificateTypes.All` or `CertificateTypes.SelfSigned`.
 
+### ChainTrustValidationMode
+
+Default value: `X509ChainTrustMode.System`
+
+The certificate presented by the client must chain to a trusted root certificate. This check allows you to control which trust store contains these root certificates.
+
+By default, the handler uses the system trust store. If the presented client certificate needs to chain to a root certificate which doesn't appear in the system trust store, this option can be set to `X509ChainTrustMode.CustomRootTrust` to make the handler use the `CustomTrustStore` instead.
```
Comment (Rick-Anderson) [inline aspnetcore/security/authentication/certauth.md:61]:
```suggestion
By default, the handler uses the system trust store. If the presented client certificate needs to chain to a root certificate which doesn't appear in the system trust store, this option can be set to [X509ChainTrustMode.CustomRootTrust](xref:System.Security.Cryptography.X509Certificates.X509ChainTrustMode.CustomRootTrust) to make the handler use the `CustomTrustStore` instead.
```

**Review by Rick-Anderson — APPROVED** [review]

File: aspnetcore/security/authentication/certauth.md [hunk 5]
```diff
@@ -52,6 +52,22 @@ Default value: `CertificateTypes.Chained`
...
+
+The certificate presented by the client must chain to a trusted root certificate. This check allows you to control which trust store contains these root certificates.
+
+By default, the handler uses the system trust store. If the presented client certificate needs to chain to a root certificate which doesn't appear in the system trust store, this option can be set to [X509ChainTrustMode.CustomRootTrust](xref:System.Security.Cryptography.X509Certificates.X509ChainTrustMode.CustomRootTrust) to make the handler use the `CustomTrustStore` instead.
+
+### CustomTrustStore
+
+Default value: Empty <xref:System.Security.Cryptography.X509Certificates.X509Certificate2Collection>
+
+If the handler's `ChainTrustValidationMode` property is set to `X509ChainTrustMode.CustomRootTrust`, this `X509Certificate2Collection` contains every certificate which will be used to validate the client certificate up to a trusted root (including the trusted root itself.)
```
Comment (Rick-Anderson) [inline aspnetcore/security/authentication/certauth.md:67]:
```suggestion
If the handler's <xref:Microsoft.AspNetCore.Authentication.Certificate.CertificateAuthenticationOptions.ChainTrustValidationMode> property is set to `X509ChainTrustMode.CustomRootTrust`, this `X509Certificate2Collection` contains every certificate which will be used to validate the client certificate up to a trusted root (including the trusted root itself.)
```

File: aspnetcore/security/authentication/certauth.md [hunk 6]
```diff
@@ -52,6 +52,22 @@ Default value: `CertificateTypes.Chained`
 
 This check validates that only the appropriate certificate type is allowed. If the app is using self-signed certificates, this option needs to be set to `CertificateTypes.All` or `CertificateTypes.SelfSigned`.
 
+### ChainTrustValidationMode
+
+Default value: [X509ChainTrustMode.System](xref:System.Security.Cryptography.X509Certificates.X509ChainTrustMode.System)
+
+The certificate presented by the client must chain to a trusted root certificate. This check allows you to control which trust store contains these root certificates.
```
Comment (Rick-Anderson) [inline aspnetcore/security/authentication/certauth.md:59]:
```suggestion
The certificate presented by the client must chain to a trusted root certificate. This check controls which trust store contains these root certificates.
```

File: aspnetcore/security/authentication/certauth.md [hunk 7]
```diff
@@ -52,6 +52,22 @@ Default value: `CertificateTypes.Chained`
...
+
+By default, the handler uses the system trust store. If the presented client certificate needs to chain to a root certificate which doesn't appear in the system trust store, this option can be set to [X509ChainTrustMode.CustomRootTrust](xref:System.Security.Cryptography.X509Certificates.X509ChainTrustMode.CustomRootTrust) to make the handler use the `CustomTrustStore` instead.
```
Comment (Rick-Anderson) [inline aspnetcore/security/authentication/certauth.md:61]:
```suggestion
By default, the handler uses the system trust store. If the presented client certificate needs to chain to a root certificate which doesn't appear in the system trust store, this option can be set to [X509ChainTrustMode.CustomRootTrust](xref:System.Security.Cryptography.X509Certificates.X509ChainTrustMode.CustomRootTrust) to make the handler use the `CustomTrustStore`.
```

File: aspnetcore/security/authentication/certauth.md [hunk 8]
```diff
@@ -52,6 +52,22 @@ Default value: `CertificateTypes.Chained`
...
+
+### CustomTrustStore
+
+Default value: Empty <xref:System.Security.Cryptography.X509Certificates.X509Certificate2Collection>
+
+If the handler's <xref:Microsoft.AspNetCore.Authentication.Certificate.CertificateAuthenticationOptions.ChainTrustValidationMode> property is set to `X509ChainTrustMode.CustomRootTrust`, this <xref:System.Security.Cryptography.X509Certificates.X509Certificate2Collection> contains every certificate which will be used to validate the client certificate up to a trusted root (including the trusted root itself.)
```
Comment (Rick-Anderson) [inline aspnetcore/security/authentication/certauth.md:67]:
```suggestion
If the handler's <xref:Microsoft.AspNetCore.Authentication.Certificate.CertificateAuthenticationOptions.ChainTrustValidationMode> property is set to `X509ChainTrustMode.CustomRootTrust`, this <xref:System.Security.Cryptography.X509Certificates.X509Certificate2Collection> contains every certificate which will be used to validate the client certificate up to a trusted root, including the trusted root.
```

**Comment by Rick-Anderson** [comment]
@edwardneal  wow, this was super helpful. Can you check my changes online (should be live in 15 minutes). [chaintrustvalidationmode](https://learn.microsoft.com/aspnet/core/security/authentication/certauth?#chaintrustvalidationmode)

**Comment by edwardneal** [comment]
@halter73 @Rick-Anderson Thanks - I didn't know about xrefs! The other adjustments make sense too. I can see the changes are live now, so all looks good to me.

**Pull request merged**